Improve flag rendering performance, also without flags

Robyt3 opened the pull request:
Cache the current and previous `CNetObj_Flag` snap objects in the `CSnapState`, instead of iterating over all snap items in the current snapshot each time that items including flags are being rendered.

Also cache the previous `CNetObj_GameData` snap object in the `CSnapState` in addition to the current one, to avoid searching for this snap item multiple times. Thereby, the `m_GameDataSnapId` member is not necessary anymore.

On the map dm1 with 64 players at zoom 0, this decreases the average frametime from 1431 µs to 1180 µs. The CPU time spent in the `CItems::OnRender` function in this scenario is reduced from 9.14% to 1.02% of the total CPU time.

Frametime comparison:
- <img src="https://github.com/user-attachments/assets/20597997-ee94-4754-a3b4-d6d0d0e03445" />
- <img src="https://github.com/user-attachments/assets/52e74098-fac5-4ad8-9bc0-4805beefefd9" />

Profiling comparison:
- Before: <img src="https://github.com/user-attachments/assets/d602031f-9608-474b-beb6-56db80d923e4" />
- After: <img src="https://github.com/user-attachments/assets/2140e7d6-0d3e-404c-8d1e-be787c68b6f8" />

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

ChillerDragon commented on src/game/client/gameclient.h:
```diff
@@ -329,9 +329,10 @@ class CGameClient : public IGameClient
 		const CNetObj_SpectatorInfo *m_pSpectatorInfo;
 		const CNetObj_SpectatorInfo *m_pPrevSpectatorInfo;
 		const CNetObj_Flag *m_apFlags[2];
+		const CNetObj_Flag *m_apPrevFlags[2];
```
While at it we could go with NUM_TEAMS where instead of 2

ChillerDragon replied:
Ah never mind ddnet broke that. Only `protocol7::NUM_TEAMS` would work and thats weird. Ignore what I said.

Robyt3 commented:
Can't use `std::vector` inside `CSnapState` because we `mem_zero` it, so I'll probably have to refactor that first.

Robyt3 commented:
Updated to use an array to support up to 1024 flags and removed the flag snap object ID validation, so mods can continue to use multiple flags. Performance without flags seems unchanged compared to previous profiling with a maximum of 2 flags.

ChillerDragon commented:
Tested on another laptop which was at 200-400 fps depending on map and zoom and its fps were also not affected at all by this pr. But the fps TRIPPLE I had on the weak laptop is huuge!!! And I guess you also measured some improvements @Robyt3. Which means so far no fps decrease was measured only increases :rocket: :rocket: :rocket:

Looks like big performance offenders were also the snap lookups. Makes me wanna push #8037 again.

Thanks @Robyt3 this is good stuff!
